src/others/translation/lv_translation.c

```diff
@@ -0,0 +1,275 @@
+                    size_t i;
+                    for(i = 0; i < trans_cnt; i++) {
+                        lv_translation_tag_dsc_t * tag_dsc = lv_array_at(&pack->translation_array, i);
+                        if(lv_streq(tag_dsc->tag, tag)) {
+                            const char * tr = tag_dsc->translations[lang];
+                            if(tr) return tr; /*Found directly*/
+
+                            tr = tag_dsc->translations[0];
+                            if(tr) {
+                                /*Use the first translation as fall back*/
+                                LV_LOG_WARN("`%s` tag is not found for `%s` language. Using `%s` instead.",
+                                            tag, selected_lang, pack->languages[0]);
+                                return tr;
+                            }
+
+                            LV_LOG_WARN("`%s` tag is not found. Using the tag as translation.", tag);
+                            return tag; /*Return the tag as a worst case option*/
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    if(lang_found) {
+        LV_LOG_WARN("`%s` tag is not found, using the tag as translation.", tag);
+    }
+    else {
+        LV_LOG_WARN("`%s` language is not found, using the `%s` as translation.", selected_lang, tag);
+    }
+
+    return tag;
+}
+
+lv_result_t lv_translation_add_language(lv_translation_pack_t * pack, const char * lang)
+{
+    if(pack->is_static) {
+        LV_LOG_WARN("Can't add language `%s` to static translation pack `%p`", lang, (void *)pack);
+        return LV_RESULT_INVALID;
+    }
+
+    pack->language_cnt++;
+    pack->languages = lv_realloc(pack->languages, sizeof(const char *) * pack->language_cnt);
+    LV_ASSERT_MALLOC(pack->languages);
+    if(pack->languages == NULL) {
+        LV_LOG_WARN("Couldn't allocate languages in `%p`", (void *)pack);
+        return LV_RESULT_INVALID;
+    }
+
+    pack->languages[pack->language_cnt - 1] = lv_strdup(lang);
+    LV_ASSERT_MALLOC(pack->languages[pack->language_cnt - 1]);
+    if(pack->languages[pack->language_cnt - 1] == NULL) {
+        LV_LOG_WARN("Couldn't allocate the new language in `%p`", (void *)pack);
+        return LV_RESULT_INVALID;
+    }
+
+    return LV_RESULT_OK;
+}
+
+
+lv_translation_tag_dsc_t * lv_translation_add_tag(lv_translation_pack_t * pack, const char * tag_name)
+{
+    if(pack->is_static) {
+        LV_LOG_WARN("Can't add tag `%s` to static translation pack `%p`", tag_name, (void *)pack);
+        return LV_RESULT_INVALID;
+    }
+
+    lv_translation_tag_dsc_t tag;
+    tag.tag = lv_strdup(tag_name);
+    LV_ASSERT_MALLOC(tag.tag);
+    tag.translations = lv_zalloc(pack->language_cnt * sizeof(const char *));
+    LV_ASSERT_MALLOC(tag.translations);
+
+    if(tag.tag == NULL || tag.translations == NULL) {
+        LV_LOG_WARN("Couldn't allocate memory for the tag's data in `%p`", (void *)pack);
+        return NULL;
+    }
```

Comment:
One of them may not be `NULL`.

```suggestion
    if(tag.tag == NULL || tag.translations == NULL) {
        LV_LOG_WARN("Couldn't allocate memory for the tag's data in `%p`", (void *)pack);
        lv_free((void *)tag.tag);
        lv_free((void *)tag.translations);
        return NULL;
    }
```